### PR TITLE
native: clear LastGasPerVote when voting for NULL

### LINKED ISSUE #3345: Clear LastGasPerVote when voting for NULL
Port https://github.com/neo-project/neo/pull/3173.

### PR DESCRIPTION
**Pull request by AliceInHunterland**
Port neo-project/neo#3173.

Close #3345

### CODE REVIEW

File: pkg/core/native/native_neo.go [hunk 1]
```diff
@@ -904,6 +904,8 @@ func (n *NEO) VoteInternal(ic *interop.Context, h util.Uint160, pub *keys.Public
 		if !cd.Registered {
 			return errors.New("validator must be registered")
 		}
+	} else {
+		acc.LastGasPerVote = *big.NewInt(0)
```
Comment (roman-khimov) [inline pkg/core/native/native_neo.go:908]: I think you're doing it after `PutStorageItem` now which defeats the purpose. We need a test for it.

Comment (AnnaShaleva) [reply pkg/core/native/native_neo.go:908]: For this particular case we should not update last gas per vote before line 894 (`PutStorageItem`). Because we have L894 to match C#'s `GetAndChange` behaviour which always saves changed storage item even if `false` is returned in the middle of `vote` handling due to some error. It affects dumps.

What we need is to place this change exactly to the place where C# code has it, to the end of the method, right before the notification emission, so that LastGasPerVote update becomes the last change of account's state before saving. This allows to avoid bugs like c63528db fixes.

Comment (AnnaShaleva) [reply pkg/core/native/native_neo.go:908]: > We need a test for it.

We need two tests, if possible: one test that checks the "successful" case of voting for null (when no error is returned from `VoteInternal`, then account's LastGasPerVote should be 0 in the end). And another one that checks that in case of early return from `VoteInternal` due to some error before L935 account's LastGasPerVote stays unchanged.

File: pkg/core/native/native_neo.go [hunk 2]
```diff
@@ -932,6 +932,9 @@ func (n *NEO) VoteInternal(ic *interop.Context, h util.Uint160, pub *keys.Public
 	}
 	ic.DAO.PutStorageItem(n.ID, key, acc.Bytes(ic.DAO.GetItemCtx()))
 
+	if pub == nil {
+		acc.LastGasPerVote = *big.NewInt(0)
+	}
```
Comment (AnnaShaleva) [inline pkg/core/native/native_neo.go:937]: You should put this right before line 933 because `ic.DAO.PutStorageItem` is exactly the method that puts the updated account state into contract storage.

File: pkg/core/native/native_test/neo_test.go [hunk 3]
```diff
@@ -118,6 +118,47 @@ func TestNEO_CandidateEvents(t *testing.T) {
...
+		err = as.FromStackItem(stack.Pop().Item())
+		require.NoError(t, err)
+		return as
+	}
+	advanceChain := func(t *testing.T) {
+		for i := 0; i < freq; i++ {
+			neoCommitteeInvoker.AddNewBlock(t)
+		}
+	}
+	txes := make([]*transaction.Transaction, 0, committeeSize*4-2)
```
Comment (AnnaShaleva) [inline pkg/core/native/native_test/neo_test.go:144]: Capacity is too big.

File: pkg/core/native/native_test/neo_test.go [hunk 4]
```diff
@@ -118,6 +118,47 @@ func TestNEO_CandidateEvents(t *testing.T) {
...
+	transferTx := neoValidatorsInvoker.PrepareInvoke(t, "transfer", e.Validator.ScriptHash(), candidate.(neotest.SingleSigner).Account().PrivateKey().GetScriptHash(), int64(committeeSize+1)*1000000, nil)
+	txes = append(txes, transferTx)
+	registerTx := neoValidatorsInvoker.WithSigners(candidate).PrepareInvoke(t, "registerCandidate", candidate.(neotest.SingleSigner).Account().PublicKey().Bytes())
+	txes = append(txes, registerTx)
+	voteTx := neoValidatorsInvoker.WithSigners(voter).PrepareInvoke(t, "vote", voter.(neotest.SingleSigner).Account().PrivateKey().GetScriptHash(), nil)
```
Comment (AnnaShaleva) [inline pkg/core/native/native_test/neo_test.go:149]: Should firstly vote for a candidate, then check the expected account state, then vote for nil.

File: pkg/core/native/native_test/neo_test.go [hunk 5]
```diff
@@ -118,6 +118,47 @@ func TestNEO_CandidateEvents(t *testing.T) {
 	require.Equal(t, 0, len(aer.Events))
 }
 
+func TestNEO_VoteForNull(t *testing.T) {
```
Comment (AnnaShaleva) [inline pkg/core/native/native_test/neo_test.go:121]: OK, you need to register and vote for new committee to properly update voters reward. This test becomes very similar to `TestNEO_Vote`, thus let's get back to the initial plan and extend `TestNEO_Vote` instead of creating a new test. In `TestNEO_Vote` you already have new committee elected, so the task is in the end of `TestNEO_Vote` to take one voter that has voted for the new committee member, check its account state, then perform unvoting andthen check its account state one more time.

Comment (AnnaShaleva) [reply pkg/core/native/native_test/neo_test.go:121]: The only thing is that you have to re-register candidate 0 back and wait for committee update one more time, because test ends with rolling back to standby validators:
https://github.com/nspcc-dev/neo-go/blob/1679b162b8d33695f63c56a472db2c98b36e26e6/pkg/core/native/native_test/neo_test.go#L264-L265

File: pkg/core/native/native_test/neo_test.go [hunk 6]
```diff
@@ -271,6 +271,28 @@ func TestNEO_Vote(t *testing.T) {
...
+		stateBeforeUnvote := getAccountState(t, voters[0].ScriptHash())
+		require.NotEqual(t, uint64(0), stateBeforeUnvote.LastGasPerVote.Uint64())
+		// unvote
+		unvoteTx := neoValidatorsInvoker.WithSigners(voters[0]).PrepareInvoke(t, "vote", voters[0].(neotest.SingleSigner).Account().PrivateKey().GetScriptHash(), nil)
+		neoValidatorsInvoker.AddNewBlock(t, unvoteTx)
+		e.CheckHalt(t, unvoteTx.Hash(), stackitem.Make(true))
+		advanceChain(t)
+
+		stateAfterUnvote := getAccountState(t, voters[0].ScriptHash())
+		fmt.Println(stateAfterUnvote.LastGasPerVote)
```
Comment (AnnaShaleva) [inline pkg/core/native/native_test/neo_test.go:293]: Not needed.

File: pkg/core/native/native_test/neo_test.go [hunk 7]
```diff
@@ -271,6 +271,33 @@ func TestNEO_Vote(t *testing.T) {
...
+
+		stateBeforeUnvote := getAccountState(t, voters[0].ScriptHash())
+		require.NotEqual(t, uint64(0), stateBeforeUnvote.LastGasPerVote.Uint64())
+		// unvote
+		unvoteTx := neoValidatorsInvoker.WithSigners(voters[0]).PrepareInvoke(t, "vote", voters[0].(neotest.SingleSigner).Account().PrivateKey().GetScriptHash(), nil)
+		neoValidatorsInvoker.AddNewBlock(t, unvoteTx)
+		e.CheckHalt(t, unvoteTx.Hash(), stackitem.Make(true))
+		advanceChain(t)
+
+		stateAfterUnvote := getAccountState(t, voters[0].ScriptHash())
```
Comment (AnnaShaleva) [inline pkg/core/native/native_test/neo_test.go:297]: `getAccountState` is not enough, you also need to ensure that resulting stackitem contains 4 subitems, and the fourth one is zero integer, because `(s *NEOBalance) FromStackItem` is able to handle both 3 and 4 subitems.

File: pkg/core/native/native_test/neo_test.go [hunk 8]
```diff
@@ -271,6 +271,35 @@ func TestNEO_Vote(t *testing.T) {
...
+	//LastGasPerVote should be 0 after unvoting
+	getAccountState := func(t *testing.T, account util.Uint160) (*state.NEOBalance, uint64) {
+		stack, err := neoCommitteeInvoker.TestInvoke(t, "getAccountState", account)
+		require.NoError(t, err)
+		res := stack.Pop().Item()
+		require.Equal(t, 4, len(res.Value().([]stackitem.Item)))
+		as := new(state.NEOBalance)
+		err = as.FromStackItem(res)
+		require.NoError(t, err)
+		return as, res.Value().([]stackitem.Item)[3].Value().(*big.Int).Uint64()
```
Comment (AnnaShaleva) [inline pkg/core/native/native_test/neo_test.go:283]: Second return value is not needed, it's enough that you've checked `require.Equal(t, 4, len(res.Value().([]stackitem.Item)))`.

File: pkg/core/native/native_test/neo_test.go [hunk 9]
```diff
@@ -271,6 +271,35 @@ func TestNEO_Vote(t *testing.T) {
 		require.NotEqual(t, candidates[0], pubs[i])
 		require.NotEqual(t, candidates[len(candidates)-1], pubs[i])
 	}
+	//LastGasPerVote should be 0 after unvoting
+	getAccountState := func(t *testing.T, account util.Uint160) (*state.NEOBalance, uint64) {
+		stack, err := neoCommitteeInvoker.TestInvoke(t, "getAccountState", account)
+		require.NoError(t, err)
+		res := stack.Pop().Item()
+		require.Equal(t, 4, len(res.Value().([]stackitem.Item)))
```
Comment (AnnaShaleva) [inline pkg/core/native/native_test/neo_test.go:279]: This line deserves a comment.

File: pkg/core/native/native_test/neo_test.go [hunk 10]
```diff
@@ -271,6 +271,35 @@ func TestNEO_Vote(t *testing.T) {
 		require.NotEqual(t, candidates[0], pubs[i])
 		require.NotEqual(t, candidates[len(candidates)-1], pubs[i])
 	}
+	//LastGasPerVote should be 0 after unvoting
```
Comment (AnnaShaleva) [inline pkg/core/native/native_test/neo_test.go:274]: s/`//LastGasPerVote should be 0 after unvoting`/`// LastGasPerVote should be 0 after unvoting.`

**Review by AnnaShaleva — COMMENTED** [review]
Otherwise LGTM, the second test is not needed, since it'll require nontrivial modifications to cause error in the middle of `voteInternal`, just doesn't worth it.

File: pkg/core/native/native_test/neo_test.go (hunk 8 above)
Comment (AliceInHunterland) [reply pkg/core/native/native_test/neo_test.go:283]: but the `fourth one is zero integer` no need to check it?

**Review by AnnaShaleva — APPROVED** [review]

File: pkg/core/native/native_test/neo_test.go (hunk 8 above)
Comment (AnnaShaleva) [reply pkg/core/native/native_test/neo_test.go]: https://github.com/nspcc-dev/neo-go/blob/9c83beffbb69a8d8a455050cfbba6aeb62d9676b/pkg/core/state/native_state.go#L162-L168
